cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -391,6 +394,35 @@ public RunId run(ProgramId programId, Map<String, String> overrides, boolean deb
     return runInternal(programId, userArgs, sysArgs, debug);
   }
 
+  /**
+   * Starts programs which were suspended between startTime and endTime.
+   *
+   * @param applicationId the application to restart programs in
+   * @param startTimeSeconds earliest stop time in seconds of programs to restart
+   * @param endTimeSeconds latest stop time in seconds of programs to restart
+   * @return {@link Set<RunId>}
+   * @throws ConflictException if the specified program is already running, and if concurrent runs are not allowed
+   * @throws NotFoundException if the specified program or the app it belongs to is not found in the specified namespace
+   * @throws IOException if there is an error starting the program
+   * @throws UnauthorizedException if the logged in user is not authorized to start the program. To start a program,
+   *                               a user requires {@link Action#EXECUTE} on the program
+   * @throws Exception if there were other exceptions checking if the current user is authorized to start the program
+   */
+  public Set<RunId> restart(ApplicationId applicationId, long startTimeSeconds, long endTimeSeconds) throws Exception {
+    Set<RunId> runs = new HashSet<>();
+    Map<ProgramRunId, RunRecordMeta> runMap =
+      store.getRuns(applicationId, ProgramRunStatus.KILLED, Integer.MAX_VALUE,
+                    meta -> meta.getStopTs() >= startTimeSeconds && meta.getStopTs() < endTimeSeconds);
+
+    for (Map.Entry<ProgramRunId, RunRecordMeta> entry : runMap.entrySet()) {
```

Comment:
Seems like you are only using the keys in the map. You can simplify the code to:

```
runs.keySet().stream()
  .map(ProgramRunId::getParent)
  .forEach(p -> runs.add(run(p, getRuntimeArgs(p), false)));
```

Reply:
Changed to keyset, but the stream is troublesome since getRuntimeArgs throws a checked exception, which we would need to wrap to a runtime exception in a separate function if we want to stream it... seems like it's not worth it